x/stake/handler.go

```diff
@@ -53,212 +63,245 @@ func EndBlocker(ctx sdk.Context, k Keeper) (ValidatorUpdates []abci.Validator) {
 // These functions assume everything has been authenticated,
 // now we just perform action and save
 
-func handleMsgCreateValidator(ctx sdk.Context, msg MsgCreateValidator, k Keeper) sdk.Result {
+func handleMsgCreateValidator(ctx sdk.Context, msg types.MsgCreateValidator, k keeper.Keeper) sdk.Result {
 
 	// check to see if the pubkey or sender has been registered before
 	_, found := k.GetValidator(ctx, msg.ValidatorAddr)
 	if found {
-		return ErrValidatorExistsAddr(k.codespace).Result()
-	}
-	if msg.Bond.Denom != k.GetParams(ctx).BondDenom {
-		return ErrBadBondingDenom(k.codespace).Result()
+		return ErrValidatorAlreadyExists(k.Codespace()).Result()
 	}
-	if ctx.IsCheckTx() {
-		return sdk.Result{}
+	if msg.SelfDelegation.Denom != k.GetParams(ctx).BondDenom {
+		return ErrBadDenom(k.Codespace()).Result()
 	}
 
 	validator := NewValidator(msg.ValidatorAddr, msg.PubKey, msg.Description)
-	k.setValidator(ctx, validator)
-	k.setValidatorByPubKeyIndex(ctx, validator)
-	tags := sdk.NewTags(
-		"action", []byte("createValidator"),
-		"validator", msg.ValidatorAddr.Bytes(),
-		"moniker", []byte(msg.Description.Moniker),
-		"identity", []byte(msg.Description.Identity),
-	)
+	k.SetValidator(ctx, validator)
+	k.SetValidatorByPubKeyIndex(ctx, validator)
 
-	// move coins from the msg.Address account to a (self-bond) delegator account
+	// move coins from the msg.Address account to a (self-delegation) delegator account
 	// the validator account and global shares are updated within here
-	delegateTags, err := delegate(ctx, k, msg.ValidatorAddr, msg.Bond, validator)
+	_, delegation, validator, pool, err := k.Delegate(ctx, msg.ValidatorAddr, msg.SelfDelegation, validator)
+	k.SetPool(ctx, pool)
+	k.SetDelegation(ctx, delegation)
+	k.UpdateValidator(ctx, validator)
 	if err != nil {
 		return err.Result()
 	}
-	tags = tags.AppendTags(delegateTags)
+
+	tags := sdk.NewTags(
+		tags.Action, tags.ActionCreateValidator,
+		tags.DstValidator, msg.ValidatorAddr.Bytes(),
```

Comment:
Needs to be passed as `[]byte(msg.ValidatorAddr.String())` so we get the hex-encoded address.

Reply:
addressed